### PR TITLE
Plugins v3

### PR DESCRIPTION
**Pull request by catornot**
yes

 
## what changed
- nuked presence ( will be move to the discord rpc plugin )
- more exposed sq functions
- exposed dll addresses
- `g_pCVar` is exposed
- added "userdata" to plugin's async call
- added runframe to plugins

discord rpc pr : https://github.com/R2Northstar/NorthstarDiscordRPC/pull/10
mods pr : https://github.com/R2Northstar/NorthstarMods/pull/652
test plugins : https://github.com/R2Northstar/NorthstarLauncher/pull/472#issuecomment-1661401264

### CODE REVIEW

**Comment by Jan200101** [comment]
Small wish I have:
make `plugin_abi.h` a pure C header.

Right now, plugins need to maintain the same data structure in their own codebase, making it rather tedious to maintain.
Turning `plugin_api.h` into a pure C header would allow languages that can understand C headers (e.g. C, C++, Zig) to import it directly, with little to no modification needed.

**Comment by catornot** [comment]
plugins that can be used to test plugins v3 :)
[r2rcon-rs](https://github.com/catornot/r2rcon-rs): [r2rcon_rs.zip](https://github.com/R2Northstar/NorthstarLauncher/files/12235405/r2rcon_rs.zip)
rrplug_test: [rrplug_test.zip](https://github.com/R2Northstar/NorthstarLauncher/files/12235418/rrplug_test.zip)
^ no src code

File: NorthstarDLL/plugins/plugin_abi.h
Comment (F1F7Y) [inline NorthstarDLL/plugins/plugin_abi.h:113]: Why expose this when you can run `CreateInterface` on `vstdlib` to get the same pointer

File: NorthstarDLL/plugins/pluginbackend.cpp
Comment (F1F7Y) [inline NorthstarDLL/plugins/pluginbackend.cpp:47]: why is this hardcoded

File: PLUGINS.ipch
Comment (F1F7Y) [inline PLUGINS.ipch:1]: this shouldnt be here

**Review by F1F7Y — DISMISSED** [review]
As much as some changes I like, my main problem with our current plugin system is that increasing the version breaks all existing plugins.

Ill probably draft up an issue of how i think we could do better as increasing version on every minor change doesnt sound feasible at all

**Comment by GeckoEidechse** [comment]
> As much as some changes I like, my main problem with our current plugin system is that increasing the version breaks all existing plugins.

In the case of v3 here, it's more than just a minor change, right? ^^

File: NorthstarDLL/plugins/pluginbackend.cpp
Comment (catornot) [reply NorthstarDLL/plugins/pluginbackend.cpp:47]: it was hardcoded since v2

Comment (F1F7Y) [reply NorthstarDLL/plugins/pluginbackend.cpp:47]: since you're increasing the version why not just improve upon the old system even more

Comment (catornot) [reply NorthstarDLL/plugins/pluginbackend.cpp:47]: what do you suggest then?

Comment (F1F7Y) [reply NorthstarDLL/plugins/pluginbackend.cpp:47]: Inform in the load library hooks, if you want to filter just game libs use only the ExA hook. Some other libs may be loaded through that though

File: NorthstarDLL/plugins/plugin_abi.h
Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h:113]: it's just easier to expose it in `m_sEngineData` since it already has stuff, altought interfaces for plugins could be better in some ways.

**Review by NoCatt — APPROVED** [review]
Made a Plugin that adds a squirrel function, no crashes ( made with rrplug )

File: NorthstarDLL/plugins/plugin_abi.h
Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h:113]: also I don't want to link rrplug to winapi yet

Comment (F1F7Y) [reply NorthstarDLL/plugins/plugin_abi.h:113]: "I don't want to link my personal project to winapi so i'm going to make the plugin api worse for everyone"

Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h:113]: how does it make it worse?

Comment (F1F7Y) [reply NorthstarDLL/plugins/plugin_abi.h:113]: Extra pointer in the struct for which there is an alternative to get it that's arguably better since it allows you to get any interface instead of hardcoding pointers to them.

Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h:113]: I would just keep it for this version since v2 already had hard coded stuff

Comment (Jan200101) [inline NorthstarDLL/plugins/plugin_abi.h:113]: All of the data functions should have their type information exposed, just like the init funcs does.

plugins that directly load types from NorthstarLauncher could then make reasonable decisions without having to literally copy a dozen files over.

Comment (Jan200101) [reply NorthstarDLL/plugins/plugin_abi.h:113]: (this is not the code I selected, go home github you are drunk)

**Comment by Jan200101** [comment]
Another thing:
the typedefs for `PluginInitFuncs` need to marked as extern C.
The functions are already marked as such, just the types aren't.

**Comment by F1F7Y** [comment]
Dismissed my review as i don't want to block this. Personally i see no point in working on a system the majority agrees should be replaced with a different solution.

The code as it is looks acceptable, but im not putting my name on it by approving. ( hence why i dismissed my review )

File: NorthstarDLL/plugins/plugin_abi.h
Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h:113]: I hate header files

Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h:116]: I have no idea how to make the compiler not complain about duplicate types

Comment (Jan200101) [reply NorthstarDLL/plugins/plugin_abi.h:116]: You could rename the typedefs to something else.
Or maybe this doesn't need to be done, since Titanfall 2 and N* are built using MSVC and MSVC is not ABI stable this could cause more issues than its worth.

File: HOSTSTATE.ipch
Comment (GeckoEidechse) [inline HOSTSTATE.ipch:1]: What does this file do?

Comment (catornot) [reply HOSTSTATE.ipch:1]: idk

Comment (catornot) [reply HOSTSTATE.ipch:1]: looks like https://github.com/R2Northstar/NorthstarLauncher/pull/472#discussion_r1281504576

Comment (GeckoEidechse) [reply HOSTSTATE.ipch:1]: Probably wanna remove it then :P

File: NorthstarDLL/plugins/plugin_abi.h
Comment (GeckoEidechse) [inline NorthstarDLL/plugins/plugin_abi.h:55]: Was there a reason to move these entries around? Grouping them with similar related entries maybe? ^^

File: NorthstarDLL/plugins/plugin_abi.h
Comment (GeckoEidechse) [inline NorthstarDLL/plugins/plugin_abi.h:137]: Trailing leftover comment? Could just remove it tbh

File: NorthstarDLL/plugins/plugins.cpp
Comment (GeckoEidechse) [inline NorthstarDLL/plugins/plugins.cpp:132]: Wait, I'm confused.

Looking at old version, it looks like `ABI_VERSION` held version number of plugin that was read when in reality it's defined in `NorthstarDLL/plugins/plugin_abi.h` as the plugin version of the client and NOT the current plugin.

I guess old log string was just plain wrong then, lol

File: NorthstarDLL/shared/gamepresence.cpp
Comment (GeckoEidechse) [inline NorthstarDLL/shared/gamepresence.cpp:1]: Just so that I'm up to speed, gamepresence handling was mostly moved in the plugin itself, right?

File: NorthstarDLL/squirrel/squirrel.cpp
Comment (GeckoEidechse) [inline NorthstarDLL/squirrel/squirrel.cpp:194]: Same question as above regarding re-ordering.

**Review by GeckoEidechse — COMMENTED** [review]
first pass, mostly bikeshedding for now

File: NorthstarDLL/plugins/plugins.cpp
Comment (F1F7Y) [reply NorthstarDLL/plugins/plugins.cpp:132]: There's multiple log calls in plugins that are just plain wrong

File: NorthstarDLL/squirrel/squirrel.cpp
Comment (catornot) [reply NorthstarDLL/squirrel/squirrel.cpp:194]: to match the `SquirrelFunctions` order

File: NorthstarDLL/plugins/plugins.cpp
Comment (catornot) [reply NorthstarDLL/plugins/plugins.cpp:132]: I didn't add new log calls

**Comment by Neoministein** [comment]
>As much as some changes I like, my main problem with our current plugin system is that increasing the version breaks all existing plugins.

I am not to familiar of how the plugin system is implemented but are the older versions of the plugin API still supported?
 
If v3 comes out, are v2 / v1 plugins still going to work? 

If the API is still available are there plans of deprecation at some point?

**Comment by Jan200101** [comment]
as it stands, bumping plugin versions automatically deprecates all prior ones.

The system as it stands right now is not suited for API versioning, #530 makes proposals for an interface that would allow for that.

**Comment by Neoministein** [comment]
>as it stands, bumping plugin versions automatically deprecates all prior ones.

Not really a fan of this tbh. 

I hoped when I created the websocket plugin that this would be a one a done situation, since I am not that big of a fan of low level programming. Hopping that the [rrplug](https://github.com/R2NorthstarTools/rrplug) wrapper makes it easier to update.


> https://github.com/R2Northstar/NorthstarLauncher/issues/530 makes proposals for an interface that would allow for that.

That would be great, but I am guessing that it will take a lot of time?

**Comment by GeckoEidechse** [comment]
> Not really a fan of this tbh.
> 
> I hoped when I created the websocket plugin that this would be a one a done situation, since I am not that big of a fan of low level programming.

Yeah kinda sucks but at the same time we really don't have the dev power to keep backwards compatibility and everything while also trying to implement other features... :/
Backwards compatibility usually brings with it a huge maintenance burden that we kinda cannot afford with everything else going on.

> Hopping that the [rrplug](https://github.com/R2NorthstarTools/rrplug) wrapper makes it easier to update.

That's the goal I think ^^

**Comment by Neoministein** [comment]
> we really don't have the dev power to keep backwards compatibility and everything while also trying to implement other features... :/

Yeah understandable, this is a hobby project for all of us I would most likely guess.

Is there a plan of having like a grace period to let plugin developers update their plugin after v3 is feature completed before it gets released in a new Northstar version?

**Comment by GeckoEidechse** [comment]
> Is there a plan of having like a grace period to let plugin developers update their plugin after v3 is feature completed before it gets released in a new Northstar version?

We will definitely send out an an announcement in modding or dev announcement channel post-merge and pre-release and also NOT version gate the release for a few weeks so that old version can still be used as fallback.

**Comment by GeckoEidechse** [comment]
#581 caused a merge conflict. It's a small change so should be quite fast to address, could probably even just merge up to the commit before it without conflicts and then merge 98af0dcbb45972d4af22ea0d26a3a2c518ea9b13 with `--strategy=ours` to just ignore the changes from it (assuming v3 does init properly).

File: NorthstarDLL/plugins/plugin_abi.h
Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h]: resolved by 6195047

**Review by Jan200101 — CHANGES_REQUESTED** [review]
I tried testing with the given plugin but it just panics and gets the game stuck after the company logos
```
[2023-10-21] [08:35:45] [PLUGINSYS] [info] Succesfully loaded R2Northstar\plugins\r2rcon_rs.dll
[2023-10-21] [08:35:45] [PLUGINSYS] [info] Loading plugin r2rcon-rs version 1.0
[2023-10-21] [08:35:45] [r2rcon-rs] [info] plugin logging initialized
[2023-10-21] [08:35:45] [r2rcon-rs] [info] plugin static initialized : true
[2023-10-21] [08:35:45] [r2rcon-rs] [error] 
[2023-10-21] [08:35:45] [r2rcon-rs] [error] plugin panicked at src\lib.rs:50:18
[2023-10-21] [08:35:45] [r2rcon-rs] [error] full message:
[2023-10-21] [08:35:45] [r2rcon-rs] [error] panicked at 'a rcon cmd wasn't present', src\lib.rs:50:18
[2023-10-21] [08:35:45] [r2rcon-rs] [error] 
```

I updated SouthRPC to Plugins v3 but calling `Cbuf_Execute` causes a crash.
Had no luck debugging it thus far.

[SouthRPC V3 Source Code](https://github.com/Jan200101/SouthRPC/tree/cat-v3)
[SouthRPC V3 Plugin Archive](https://github.com/R2Northstar/NorthstarLauncher/files/13060320/SouthRPC.zip)

You can test SouthRPC is working by sending a request to
`http://localhost:26503`
with the body
```
{
	"jsonrpc": "2.0",
	"method": "execute_squirrel",
	"params": {
		"code": "printt(\"Hello World!\")"
	}
}
```
Here is a short form curl command:
`curl --request POST --url http://localhost:26503/ --header 'Content-Type: application/json' --data '{"jsonrpc": "2.0", "method": "execute_squirrel", "params": {"code": "printt(\"Hello World!\")"}}'`

File: NorthstarDLL/plugins/plugin_abi.h
Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h]: resolved by [4c4cb1b](https://github.com/R2Northstar/NorthstarLauncher/pull/472/commits/4c4cb1b8e9d7c24a6cb07a3c9f6fb82aa9424216)

File: NorthstarDLL/plugins/plugin_abi.h
Comment (catornot) [reply NorthstarDLL/plugins/plugin_abi.h]: resolved by [4c4cb1b](https://github.com/R2Northstar/NorthstarLauncher/pull/472/commits/4c4cb1b8e9d7c24a6cb07a3c9f6fb82aa9424216)

**Comment by catornot** [comment]
> I tried testing with the given plugin but it just panics and gets the game stuck after the company logos

panicking is intended behavior since it didn't find the correct command line args and it's in a thread so it's fine.
The game probably got stuck because that plugin is outdated and shouldn't work after [b292371](https://github.com/R2Northstar/NorthstarLauncher/pull/472/commits/b292371f95405fa1ac0255cd3939d347775b0e65).

**Review by Jan200101 — APPROVED** [review]
Tested already with SouthRPC.
Code is good.

**Review by GeckoEidechse — APPROVED** [review]
Confirmed working in testing together with https://github.com/R2Northstar/NorthstarMods/pull/652 and https://github.com/R2Northstar/NorthstarDiscordRPC/pull/10

Joined a bunch of servers and watched Discord activity status update

File: NorthstarDLL/plugins/plugins.cpp
Comment (catornot) [reply NorthstarDLL/plugins/plugins.cpp:132]: should this get resolved or not?

**Review by Jan200101 — APPROVED** [review]
Looked over the code again and it still looks good to me.

File: NorthstarDLL/plugins/plugins.cpp
Comment (GeckoEidechse) [reply NorthstarDLL/plugins/plugins.cpp:132]: I moved the change to #591 which I'll merge first and then resolve the conversation here ^^

**Comment by GeckoEidechse** [comment]
merged based on reviews ig

**Pull request merged**
